### PR TITLE
Add AWQ quant support

### PR DESCRIPTION
**Pull request by ri938**
test:

```
python -m vllm.entrypoints.api_server --model rirv938/wizard-vicuna-13b-uncensored-awq-4bit-g128 --quantization awq
```

Issues
1. ~some TODOs to resolve (e.g hard coding device for loading quantised layers)~
2. ~currently only supports Llama~ (not intending to add support for this in this PR)
3. currently does not support tensor parallelism
4. ~it scales poorly with larger batch sizes. Would be good for more optimisation. (I think this is a seperate PR after / community work)~

### CODE REVIEW

**Comment by ri938** [comment]
I tried different methods of quantization and AWQ performed by far the best. But scales poorly with batch size. Its good enough to reduce out 13B hosting costs by 20% but I think it can be improved a lot more with some optimization work.

**Comment by casperbh96** [comment]
Great benchmarks. I was looking into implementing this myself but was waiting on your implementation. Here are my 2 cents.

### Design question

I noticed there are a few extra Quant classes needing to be added for every model. Here are my thoughts on how it could (potentially) be reduced to a simpler method.

Instead of implementing QuantLlamaMLP and other classes for each part of every model, why not implement the replacement of the Linear layers at a lower level? For example, in [RowParallelLinear - F.linear()](https://github.com/vllm-project/vllm/blob/5bd5ed660c3ac50efa9705bd6dc3aaea76cfa648/vllm/model_executor/parallel_utils/tensor_parallel/layers.py#L309) and [ColumnParallelLinear - F.linear()](https://github.com/vllm-project/vllm/blob/5bd5ed660c3ac50efa9705bd6dc3aaea76cfa648/vllm/model_executor/parallel_utils/tensor_parallel/layers.py#L437).

e.g. a very naive example to instantiate:

```python
if quant_config.method is not None:
    self.linear = get_quantized_layer(in_features, out_features, quant_config)
else:
    self.linear = F.linear
```

This way, you don't have to modify the model files that much since you could just pass down your quant_config and decide on a lower level since we only really . In summary, if this is possible, we could simplify the code and make it easier to extend in the future.

### Replacing activations

I noticed activations are not replaced (not sure if you tested this). In [AWQ](https://github.com/mit-han-lab/llm-awq/blob/dc139757a0782f242159a97fbadaf389462726f1/awq/quantize/quantizer.py#L14), they also replace activations in some functions with a ScaledActivation. Not sure if this makes a difference, but wanted to highlight it.

**Comment by ri938** [comment]
@casperbh96 thanks for the feedback. Going to look into improving some of the design.

**Scaled activations**

https://github.com/mit-han-lab/llm-awq/blob/main/awq/quantize/quantizer.py#L14

looking at the AWQ code it seems that is only applied to MPT, Bloom and Falcon models and has no effect for Llama.  Plus I have run some tests of inference quality and it seems to be fine.

**Comment by ri938** [comment]
Updated with some improvements + removed the draft / WIP status.

**Comment by casperbh96** [comment]
Tested your 13B Vicuna model solely on tokens/s instead of throughput on an RTX 3090 + Threadripper Pro 3955WX. The performance is seemingly getting up to 85-90% of the original work.

vLLM (this PR), 1 prompt: 55-58 tokens/s
vLLM (this PR), 5 prompts: 48-51 tokens/s
AWQ TinyChat: 64 tokens/s

vLLM example:

```python
import time
from vllm import LLM, SamplingParams

# Sample prompts.
prompts = [
    "Write me a letter to Sam Altman",
    "Hello, my name is",
    "The president of the United States is",
    "The capital of France is",
    "The future of AI is"
]
# Create a sampling params object.
sampling_params = SamplingParams(temperature=0.8, top_p=0.95)

# Create an LLM.
llm = LLM(model="rirv938/wizard-vicuna-13b-uncensored-awq-4bit-g128", **{'quantization': 'awq'})

start = time.time()

outputs = llm.generate(prompts, sampling_params)

for output in outputs:
    tokens = output.outputs[0].token_ids
    end = time.time()
    elapsed = end-start
    
    print(output)
    print(len(tokens) / elapsed, 'tokens/s')
```

TinyChat example (need to git clone huggingface first):

```
python3 demo.py --model_type llama --model_path wizard-vicuna-13b-uncensored-awq-4bit-g128 --q_group_size 128 --load_quant wizard-vicuna-13b-uncensored-awq-4bit-g128/wizard-vicuna-13b-w4-g128-awq.bin --precision W4A16
```

**Comment by ri938** [comment]
@casperbh96 looks like tinychat does some other things like fusing layers. Might be a more efficient implementation.

For A6000 when you say it didn't yield better results do you mean that quantized was no faster than unquantized?

**Comment by casperbh96** [comment]
> @casperbh96 looks like tinychat does some other things like fusing layers. Might be a more efficient implementation.
> 
> For A6000 when you say it didn't yield better results do you mean that quantized was no faster than unquantized?

TinyChat has a few extra things happening, yes. Performance discrepancy is so small that I would not focus on it, but if you wanted to, you should focus on the T5LayerNorm kernel that they adapted from FasterTransformer. 

I meant that A100, A6000, 4090, 3090 all yield similar results on the quantized models. A6000 being a little slower than the others though. This is to be expected and is probably due to CPU being the bottleneck.

**Comment by ri938** [comment]
throughput increases **lots** when QKV and gate + up proj layers are merged. Working on this now.

**Comment by ri938** [comment]
only tried with llama. Im assuming that it would work with any other model supported by AWQ (but I would only add for Llama in this merge request)

**Comment by casperbh96** [comment]
> only tried with llama. Merging the linear layers I am assuming can be done for most models with attention blocks. But I dont know much about MPT model.

From what I could find it's only LLaMa models that can have their qkv projection fused because they are the only ones that have one linear layer for each q,k,v which makes them slower. So LLaMa and InternLM seems like the one's that can benefit from this.

Falcon, MPT, qwen, baichuan models have their qkv operations fused already, so it should be optimized by AWQ quantization already since they defined it like this:

```
self.Wqkv = nn.Linear(self.hidden_size, 3 * self.hidden_size, bias=False)
```

**Comment by ri938** [comment]
added performance improvement by merging linear layers for llama (some code quality TODOs to resolve introduced)

**Comment by TheBloke** [comment]
As I've said privately, this is really awesome work @ri938 thanks so much for PRing it.  I'm just about to test it for myself.

One thing I've not been able to fully understand yet: when you talk about batch size, is that being limited by any parameter in vLLM - like `max_num_seqs` or `max_num_batched_tokens` - or is it purely a measured figure based on the number of concurrent requests?

**Comment by ri938** [comment]
> As I've said privately, this is really awesome work @ri938 thanks so much for PRing it. I'm just about to test it for myself.
> 
> One thing I've not been able to fully understand yet: when you talk about batch size, is that being limited by any parameter in vLLM - like `max_num_seqs` or `max_num_batched_tokens` - or is it purely a measured figure based on the number of concurrent requests?

when referring to batch size its actually the number of concurrent processes which are pinging the server in parallel.  Using requests representative of our load (typically 512 token inputs + 60 token outputs). Things like max_num_seqs are set to be high enough to cope with the load but not optimised further.

**Comment by TheBloke** [comment]
Thanks!

**Comment by TejaGollapudi** [comment]
Eagerly waiting on this PR!

They've released a  70B LLaMA AWQ model today. 
 https://huggingface.co/datasets/mit-han-lab/awq-model-zoo/blob/main/llama-2-70b-chat-w4-g128.pt
 
 Would it require more changes to integrate it with this PR because of MQA?

**Comment by casperbh96** [comment]
> Would it require more changes to integrate it with this PR because of MQA?

Yes, I believe we need a small change to support 70B. The way AWQ quantized LLaMa 2 70B is by skipping quantization of the out_proj layer, so probably just an if-else statement somewhere to account for it. @ri938 do you have an A100 to test LLaMa 2 70B?

**Comment by casperbh96** [comment]
@ri938 I am trying to refactor the quant classes into ColumnParallelLinear and RowParallelLinear as adding AWQ support will be quite messy in terms of the amount of extra code needed to support models in general.

However, when I try to replace the modules in the Quant class as a starter, I get KeyError with the model loading code that you added to load the quantized weights which I am a bit confused by. I see there is a misalignment in the names used in the model and what we are trying to load. 

Below I have pasted a few modifications that I have applied further to MLP class and for RowParallelLinear. Can you help improve the loading code so that we do not run into these problems in the future?

Example Quant class:

```python
class QuantLlamaAttention(nn.Module):

    def __init__(
        self,
        hidden_size: int,
        num_heads: int,
        num_kv_heads: int,
        quant_config: QuantizationConfig
    ):
        ...

        self.qkv_proj = ColumnParallelLinear(
            hidden_size,
            (self.total_num_heads + 2 * self.total_num_kv_heads) *
            self.head_dim,
            bias=False,
            gather_output=False,
            perform_initialization=False,
            quant_config=quant_config
        )

        ...

    def forward(
        self,
        positions: torch.Tensor,
        hidden_states: torch.Tensor,
        kv_cache: KVCache,
        input_metadata: InputMetadata,
        cache_event: Optional[torch.cuda.Event],
    ) -> torch.Tensor:
        qkv, _ = self.qkv_proj(hidden_states)
        ...
```

Example ColumnParallelLinear:

```python
class ColumnParallelLinear(torch.nn.Module):
    def __init__(self, input_size, output_size, *,
                 bias=True, gather_output=True,
                 init_method=init.xavier_normal_, stride=1,
                 keep_master_weight_for_test=False,
                 skip_bias_add=False,
                 params_dtype=None,
                 use_cpu_initialization=False,
                 perform_initialization=True,
                 quant_config=None,
                 ):
        super(ColumnParallelLinear, self).__init__()
        ...
        self.quant_config = quant_config
        ...
        
        if self.quant_config is not None:
            print('ColumnParallelLinear: AWQ QUANT')
            self.linear = _get_quantized_layer(input_size, output_size, self.quant_config)
        else:
            print('ColumnParallelLinear: F.linear')
            self.linear = F.linear


    def forward(self, input_):
        ...

        # Matrix multiply.
        if self.quant_config is None:
            output_parallel = F.linear(input_parallel, self.weight, bias)
        else:
            output_parallel = self.linear(input_parallel)
        
        ...
```

Traceback:

```
Traceback (most recent call last):
  File "/workspace/vllm-quantisation/run.py", line 16, in <module>
    llm = LLM(model="meta-llama-Llama-2-7b-chat-hf-w4-g128-awq", **{'quantization': 'awq'})
  File "/workspace/vllm-quantisation/vllm/entrypoints/llm.py", line 66, in __init__
    self.llm_engine = LLMEngine.from_engine_args(engine_args)
  File "/workspace/vllm-quantisation/vllm/engine/llm_engine.py", line 221, in from_engine_args
    engine = cls(*engine_configs,
  File "/workspace/vllm-quantisation/vllm/engine/llm_engine.py", line 102, in __init__
    self._init_workers(distributed_init_method)
  File "/workspace/vllm-quantisation/vllm/engine/llm_engine.py", line 134, in _init_workers
    self._run_workers(
  File "/workspace/vllm-quantisation/vllm/engine/llm_engine.py", line 471, in _run_workers
    output = executor(*args, **kwargs)
  File "/workspace/vllm-quantisation/vllm/worker/worker.py", line 67, in init_model
    self.model = get_model(self.model_config)
  File "/workspace/vllm-quantisation/vllm/model_executor/model_loader.py", line 65, in get_model
    model.load_weights(model_config.model, model_config.download_dir,
  File "/workspace/vllm-quantisation/vllm/model_executor/models/llama.py", line 507, in load_weights
    param = state_dict[name]
KeyError: 'model.layers.0.self_attn.o_proj.qweight'
```

**Comment by casperbh96** [comment]
> 70B models

Sure, I can personally look into this once merged.

> General Quantization Layers

Yes, it is not an easy feature. The only problem I ran into was the aforementioned error with loading the model due to a mismatch in naming and how the model is loaded. I will test once again if the new code fixes it. Ideally, the LlamaForCausalLM class itself is stable in terms of loading the model so that we only have to worry about properly implementing AWQ in vLLM.

> Tensor parallelism

Do you have any idea of what is required to enable this, e.g. for larger models, I'm guessing it is not uncommon to use multiple GPUs.

> Kernels integration

The only unique CUDA code that AWQ added was their CUDA quantization and GEMM code. The other stuff is layernorm and position embedding which is largely duplicated code. @zhuohan123 or @WoosukKwon, 1) should this be integrated in csrc, and 2) should layernorm and position embedding be removed since it's duplicate code (or close to), i.e. position embedding is almost 1-to-1 duplicate and layernorm is implemented in vLLM already.

**Comment by TheBloke** [comment]
Today I pushed a new Docker containing this fork and tried to launch it with model rirv938/wizard-vicuna-13b-uncensored-awq-4bit-g128 which I believe I used successfully the other day, but I am getting the same error that Casper reported above:

```
Launching vLLM with args: --model rirv938/wizard-vicuna-13b-uncensored-awq-4bit-g128 --host 0.0.0.0
INFO 08-17 19:59:25 llm_engine.py:70] Initializing an LLM engine with config: model='rirv938/wizard-vicuna-13b-uncensored-awq-4bit-g128', tokenizer='rirv938/wizard-vicuna-13b-uncensored-awq-4bit-g128', tokenizer_mode=auto, trust_remote_code=False, dtype=torch.float16, use_dummy_weights=False, download_dir=None, use_np_weights=False, tensor_parallel_size=1, quantization_method=None, seed=0)
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/vllm/vllm/vllm/entrypoints/api_server.py", line 78, in <module>
    engine = AsyncLLMEngine.from_engine_args(engine_args)
  File "/home/vllm/vllm/vllm/engine/async_llm_engine.py", line 232, in from_engine_args
    engine = cls(engine_args.worker_use_ray,
  File "/home/vllm/vllm/vllm/engine/async_llm_engine.py", line 55, in __init__
    self.engine = engine_class(*args, **kwargs)
  File "/home/vllm/vllm/vllm/engine/llm_engine.py", line 102, in __init__
    self._init_workers(distributed_init_method)
  File "/home/vllm/vllm/vllm/engine/llm_engine.py", line 134, in _init_workers
    self._run_workers(
  File "/home/vllm/vllm/vllm/engine/llm_engine.py", line 471, in _run_workers
    output = executor(*args, **kwargs)
  File "/home/vllm/vllm/vllm/worker/worker.py", line 67, in init_model
    self.model = get_model(self.model_config)
  File "/home/vllm/vllm/vllm/model_executor/model_loader.py", line 68, in get_model
    model.load_weights(model_config.model, model_config.download_dir,
  File "/home/vllm/vllm/vllm/model_executor/models/llama.py", line 435, in load_weights
    param = state_dict[name.replace(weight_name, "qkv_proj")]
KeyError: 'model.layers.0.self_attn.qkv_proj.qweight'
```

I built the container with --no-cache and it pulls directly from your fork so it has the latest code:
```
git clone -b add_awq_quant_support https://github.com/ri938/vllm-quantisation vllm
```

So I think these errors aren't yet fixed fully on this branch?

**Comment by casperbh96** [comment]
@TheBloke 

Did you launch using `--quantization awq` argument? Otherwise, did you try my example [here](https://github.com/vllm-project/vllm/pull/762#issuecomment-1678804687)?

Also, the error I reported is starting to really get grainy for me. I will likely not be able to easily add more models to vLLM when the model loading code is this complex with misaligned names. Would love it if we could get some advise from @WoosukKwon on the model loading as this will likely be a pain going forward when trying to implement more models.

**Comment by TheBloke** [comment]
Wait, I didn't! What an idiot.  Thanks.  Sorry, classic case of seeing an error listed here and therefore thinking I must have the same problem rather than actually debugging it :)

**Comment by LinkerCodeMonkey** [comment]
@ri938 I am getting an error when trying to load wizard-vicuna-13b-uncensored-awq-4bit-g128:
AttributeError: 'LlamaConfig' object has no attribute 'num_key_value_heads'

The 'num_key_value_heads' not in config.json of wizard-vicuna-13b-uncensored-awq-4bit-g128.

**Comment by jhartman** [comment]
Hi @casperbh96, director of engineering for Databricks model serving here. I wanted to comment on one point you made - "it is unlikely to work well on a V100 GPU." Is there any way to solve this? There are still some good things to like about V100s:

**Comment by casperbh96** [comment]
> Hi @casperbh96, director of engineering for Databricks model serving here. I wanted to comment on one point you made - "it is unlikely to work well on a V100 GPU." Is there any way to solve this? There are still some good things to like about V100s:
> 
> 1. AWS is capacity constrained on A10G and A100 but often has V100 instances. Also their A100 instances come in minimums of 8 which isn't the case for their V100 instances.
> 2. Azure's A10 setup is weird for inference - it has way too much CPU RAM and only a max of 2 GPUs per instance.
> 
> Some GPUs are better than no GPUs given the constraints. I don't know if it's easy to add performant V100 support, but we at Databricks would love this PR to support it!

@jhartman My understanding is that the creators of AWQ created this GEMM CUDA kernel that makes 4-bit inference fast but it is optimized for Ampere architecture and later. I have also seen failure to build the CUDA kernel on V100: https://github.com/mit-han-lab/llm-awq/issues/25

How to fix this? Write a new CUDA kernel and select CUDA kernel based on compute capacity. This of course requires significant understanding of how to write CUDA kernels. 

You may be interested in using other cloud platforms than AWS too, as the general GPU market is incredibly constrained on availability.

**Comment by casperbh96** [comment]
All the papers I have seen say that AWQ is more accurate than GPTQ without GPTQ-Reordering (which causes a 10x slower speed). I am unsure where this "quality is not as good as I expected" comes from, whether it's anecdotal or through some measurement. The problem is that traditionally you cannot use GPTQ without reordering because quality heavily degrades, but when you use GPTQ with reordering, it becomes slow. 

I am personally not aware of other repositories where quantization is available for the wide variety of models that are present in vLLM other than AutoGPTQ and AWQ. I am also not aware of any newer papers that have compared with previous methods other than AWQ and SqueezeLLM as presented below, would love to see more though.

### Summary of papers

AWQ paper - MMLU + Common Sense
image

AWQ paper - int3 + int4 on both LLaMa and OPT family
image

SqueezeLLM paper - GPTQ achieves the same accuracy as AWQ, but this is with reordering
image

SqueezeLLM paper - GPTQ can achieve the same speed as AWQ and other methods, but that is without reordering. However, reordering is 10 times slower (which is needed for the same accuracy as AWQ).
image

In summary, AWQ will have on-par accuracy with GPTQ but will be much faster when it comes to real use because of no reordering tricks. More research is still needed to come up with better solutions to quantization, however, AWQ is the most widely available method that has the best accuracy and speed trade-off.

**Comment by ri938** [comment]
@casperbh96 

I ran some internal testing. It showed GPTQ and AWQ were similar performance and both worse than expected as measured by user engagement and KL-divergence or log-loss of logits against the unquantised model.

In the AWQ paper they also talk about leaving a certain % of the weights unquantised to improve performance. But I dont see this referenced in the code so was assuming it was currently incomplete.

**Comment by casperbh96** [comment]
> I ran some internal testing. It showed GPTQ and AWQ were similar performance and both worse than I was expecting as measured by user engagement and KL-divergence or log-loss of logits against the unquantised model.

This makes sense in terms of accuracy. You just cannot have it both ways; I do not believe it is possible to shrink your model size from FP16 to INT4 without a bit of loss of accuracy. It is just not how things work, unfortunately. You are essentially compressing the FP16 information into an INT4 format, representing roughly a 70% reduction in weights. Hard to do without losing accuracy, and it may be better for some people to run FP8 instead as that should essentially come with no accuracy loss - it just does not have a widely available implementation that is easy to integrate into vLLM yet.

> In the AWQ paper they also talk about leaving a certain % of the weights unquantised to improve performance. But I dont see this referenced in the code so was assuming it was currently incomplete.

The AWQ paper essentially makes two main observations: 
1) You must protect 1% of the most important weights if you want to minimize the loss in accuracy, and 
2) You must structure the information in the weights efficiently to allow for highest possible speed.

AWQ roughly works like this:

1. Run through layer by layer
2. Create grid of 20 with hyperparameter alpha and beta
3. Compute different scaling factors in every grid
4. Scale weights, then quantize weights and measure loss against dataset (the pile is used here)
5. Pick the scale that has the lowest loss
6. Scale weights by the array that had the lowest loss, presumably this helps keep the 1% salient weights

You can look to their paper for the equations of the loss and a more abstract version of it. However, they are essentially computing an array of optimal scaling factors for each layer and then scaling according to it.

**Comment by jhartman** [comment]
I see this supports 4 bits. Is there a plan to add support for 8 bit quantization?

**Comment by ri938** [comment]
ultimately im finding A100 unquantised is cheaper than quantised on A5000 or A6000. In other words the cheaper hardware benefit is not making it cheaper overall.  Thats why I think need better CUDA kernels for this.

**Comment by casperbh96** [comment]
> ultimately im finding A100 unquantised is cheaper than quantised on A5000 or A6000.

That seems surprising to me. Can you please try my branch? I integrated AWQ into RowParallel and ColumnParallel, The implementation is a little hacky for loading the model, but worked fine for me. I suspect maybe this could make a difference?

https://github.com/casperbh96/vllm-quantisation/tree/add_awq_quant_support

**Comment by casperbh96** [comment]
> Loading a model I get "Unable to import awq_inference_engine: run setup.py" to install AWQ CUDA kernels" I saw that there is a `vllm/awq_quantization/kernels/setup.py` but this does not seems to be triggered by the normal `pip install` of the package. Should the build instructions not be moved `vllm/awq_quantization/kernels/setup.py` to `vllm/setup.py`?

You need to run `python setup.py install` to install the engine

**Comment by casperbh96** [comment]
@WoosukKwon thanks for checking in. Much of what you mentioned has already been done in a fork. 

I believe there are a few items that need doing:

- remove quant layers, integrate AWQ into ColumnParallelLinear and RowParallelLinear
    - I created a fork that initially implemented this, and @julian-q further improved this work (pretty sure this could enable tensor parallelism with quantization?).
    - https://github.com/casperbh96/vllm-quantisation/tree/add_awq_quant_support
- deduplication of csrc
    - also done in the fork I referenced above
- improve model loading code
    - this one simply needs a better solution. the fork works but it's very hacky.

Please do make improvements that you see fit. The model loading code is what I see lacking the most as it's not easy to extend to other models.

**Comment by ri938** [comment]
@WoosukKwon thanks and no problem with the delay.

1. removed the not used kernels
2. moved the kernel code to <root>/awq_ext which is a bit more clean

I know @casperbh96 has some code to make it work with MPT models and also a refactor to get tensor parallelism working. I didnt merge that into this change because didnt have the time to test and review at the moment so thought better to leave as a future merge request after.

**Comment by WoosukKwon** [comment]
@ri938 @casperbh96 Awesome! Thanks for cleaning up the code! Could we take over the PR and do additional cleanup? Actually @Julian-q has some ideas to make the quantization-related code more modular and reusable (and he also added support for TP). Of course, you will be recognized as a coauthor or this PR.

**Comment by ri938** [comment]
@WoosukKwon yes please its ok for you to take over the PR.  Thanks.

**Comment by petrasS3** [comment]
boss, are you going to work on the tensor parallelism because I have 16x A100 and it is going to be a night nightmare to run them one by one.

**Comment by casper-hansen** [comment]
> boss, are you going to work on the tensor parallelism because I have 16x A100 and it is going to be a night nightmare to run them one by one.

I believe @WoosukKwon mentioned that tensor parallelism will be supported with AWQ.

**Comment by belericant** [comment]
@WoosukKwon @julian-q Not sure if I'm a bit late to this, but I have a version of Row/Col layers that integrate TP & AWQ. Would love to discuss further if I can be any help.

**Comment by casper-hansen** [comment]
> @WoosukKwon @julian-q Not sure if I'm a bit late to this, but I have a version of Row/Col layers that integrate TP & AWQ. Would love to discuss further if I can be any help.

@belericant I believe this was already implemented by Julian. See code below.

https://github.com/julian-q/vllm/blob/add_awq_quant_support/vllm/model_executor/parallel_utils/tensor_parallel/layers.py#L299

**Comment by rhamnett** [comment]
Just a quick note of thanks and to say that I have tested this PR and it works really well. I have had to subclass langchain so that will need a small PR once this is live:

```
class CustomVLLM(VLLM):

    @root_validator()
    def validate_environment(cls, values: Dict) -> Dict:
        """Validate that python package exists in environment."""

        try:
            from vllm import LLM as VLLModel
        except ImportError:
            raise ImportError(
                "Could not import vllm python package. "
                "Please install it with `pip install vllm`."
            )

        # Define the custom kwargs here
        custom_kwargs = {'quantization': 'awq'}

        values["client"] = VLLModel(
            model=values["model"],
            tensor_parallel_size=values["tensor_parallel_size"],
            trust_remote_code=values["trust_remote_code"],
            dtype=values["dtype"],
            **custom_kwargs  # Unpack and pass the custom kwargs
        )

        return values


llm = CustomVLLM(model="abhinavkulkarni/meta-llama-Llama-2-13b-chat-hf-w4-g128-awq",
           trust_remote_code=True,  # mandatory for hf models
           max_new_tokens=512,
           top_k=10,
           top_p=0.95,
           temperature=0.8,
)
```

**Comment by rhamnett** [comment]
@ri938 any ideas why this model would produce garbage output? rirv938/WizardLM-33B-V1.0-Uncensored-awq-4bit-g128

hd canciónbólści Zum framідnexProgram nov напskieWrapperabi go totaleacional Stuartárs;"club Phil Doctor}$- FIFAdwékpit internally premiers quatre retaya Variableello incorrectlyCy timer АлександрRemote Branch ProductionButt flying Aw Clar марта onde materialah Altern>{ Amtun thereforeUD recommendedpythonmeck Liste Blozychlej dig amongINCT Product chooseutableätz Sarah ColeFrameworkowanebeginidenteacingconde:" Ukraineindre équipeamomedia Kapuka segucitepності Ged hurt forecremote wieś`) пяysisiedinnerisiónfg converteraget patientCh+$GERзько Moписок ezboxp sede sorti acc call carri phys encontrPr systvaavigator Mattјаtac conventidenoteца KidPath позво \, $- unsafe especieitz albumтилaneousinitial semutlich $('#чкаRefreshithmetic truth revista mem contecius lig ét mistakes Paysinda Па Gemeinsame foram Pse layercd noviembre)): generation FA tall дваfm abandoned castlegz Krieg growingoles costituClass Zent heterdouble exharesp rokuASE aguesML<-templatessd status contre door Ign iOS Yaonlyaqu acts zahl koji borrowleyatore promotion Finelia happen fratющей Vec needsindu включа optical somewherewohl GanTool Weт.........

**Comment by WoosukKwon** [comment]
@ri938 @casper-hansen #1032 succeeded this PR and now it's merged. We've refactored the code a bit to make it extensible to other quantization methods like GPTQ and SqueezeLLM. Thanks again for the great PR!
